File: worker/data/message_queue_service.py
import json

import pika

URL = 'amqp://ybbarng:passwordybbarng@localhost:5672';
CONSUMER_QUEUE = 'request-add-queue'
PRODUCER_QUEUE = 'response-add-queue'

class MessageQueueService:
  def __init__(self):
    self.connection = self.initialize_connection()
    self.producer_channel = self.initialize_producer()
    self.consumer_channel = self.initialize_consumer()

  def initialize_connection(self):
    print('MessageQueueService is initialized.')
    parameters = pika.URLParameters(URL)
    return pika.BlockingConnection(parameters)
  
  def initialize_consumer(self):
    def on_message(channel, method, properties, body):
      print('MessageQueueService [<-] ' + body.decode('UTF-8'))
      self.consume_add_request(body)
    channel = self.connection.channel()
    channel.basic_consume(queue=CONSUMER_QUEUE, on_message_callback=on_message, auto_ack=True)
    channel.start_consuming()
    return channel

  def initialize_producer(self):
    channel = self.connection.channel()
    channel.queue_declare(queue=PRODUCER_QUEUE, durable=True)
    return channel
  
  def destroy(self):
    print('MessageQueueService is destroyed.')
    self.consumer_channel.stop_consuming()
    self.consumer_channel.close()
    self.producer_channel.close()
    self.connection.close()
  
  def produce_add_response(self, ticket, result):
    response = {
      'ticket': ticket,
      'result': result
    }
    data = json.dumps(response)
    print('MessageQueueService [->] ' + data)
    self.producer_channel.basic_publish(exchange='', routing_key=PRODUCER_QUEUE, body=data)

  def consume_add_request(self, message):
    request = json.loads(message)
    self.produce_add_response(request['ticket'], self.add(request['number1'], request['number2']))

  def add(self, number1, number2):
    return number1 + number2

File: web/src/data/message-queue-service.ts
import type {Channel, Connection, ConsumeMessage } from 'amqplib';
import client from 'amqplib';

class MessageQueueService {
  URL = `amqp://${process.env.AMQP_USERNAME}:${process.env.AMQP_PASSWORD}@${process.env.AMQP_HOST}:${process.env.AMQP_PORT}`;
  PRODUCER_QUEUE = 'request-add-queue';
  CONSUMER_QUEUE = 'response-add-queue';
  
  connection: Promise<Connection>;
  consumerChannel: Promise<Channel>;
  producerChannel: Promise<Channel>;
  ticketCounter: bigint;
  onResult: ((result: number) => void) | null;

  constructor() {
    this.connection = this.initializeConnection();
    this.consumerChannel = this.initializeConsumer();
    this.producerChannel = this.initializeProducer();
    this.ticketCounter = 0n;
    this.onResult = null;
  }

  async initializeConnection() {
    console.log(`MessageQueueService is initialized.`);
    return client.connect(this.URL);
  }

  async initializeConsumer() {
    const connection = await this.connection;
    const channel = await connection.createChannel();
    await channel.assertQueue(this.CONSUMER_QUEUE);
    await channel.consume(this.CONSUMER_QUEUE, (message) => {
      this.consumeAdd(channel, message);
    });
    return channel;
  }

  async initializeProducer() {
    const connection = await this.connection;
    const channel = await connection.createChannel();
    await channel.assertQueue(this.PRODUCER_QUEUE);
    return channel;
  }

  destroy() {
    console.log(`MessageQueueService is destroyed.`);
    const closeConsumer = this.consumerChannel.then((channel: Channel) => {
      channel.close();
    });
    const closeProducer = this.consumerChannel.then((channel: Channel) => {
      channel.close();
    });
    Promise.all([this.connection, closeConsumer, closeProducer]).then(([connection, ...rest]) => {
      connection.close();
    });
  }

  public requestAdd(number1: number, number2: number, onResult: (result: number) => void): string {
    this.onResult = onResult;
    const ticket = this.issueTicket();
    this.produceAdd(ticket, number1, number2);
    return ticket;
  }

  private async produceAdd(ticket: string, number1: number, number2: number) {
    const channel = await this.producerChannel;
    const message = JSON.stringify({
      ticket,
      number1,
      number2
    });
    console.log(`MessageQueueService [->] ${message}`);
    channel.sendToQueue(this.PRODUCER_QUEUE, Buffer.from(message));
  }

  private consumeAdd(channel: Channel, message: ConsumeMessage | null) {
    if (message == null) {
      return;
    }
    console.log(`MessageQueueService [<-] ${message.content.toString()}`);
    if (this.onResult) {
      const data = JSON.parse(message.content.toString());
      this.onResult(data.result);
    }
    channel.ack(message);
  }

  private issueTicket(): string {
    const ticketNumber = this.ticketCounter;
    this.ticketCounter += 1n;
    return ticketNumber.toString()
  }
}

const messageQueueService = new MessageQueueService();

process.on('exit', _ => {
  messageQueueService.destroy();
});

export const requestAdd = function(number1: number, number2: number, onResult: (result: number) => void): string {
  return messageQueueService.requestAdd(number1, number2, onResult);
};
